### DISCUSSION TITLE
Jan Integration with Obsidian

### DISCUSSION THREAD

**Discussion by ux-han** [janhq/jan#3684] [Feature Ideas]
Requested by: https://discord.com/channels/1107178041848909847/1230524449917239337

----

Integrating Jan with Obsidian vaults and .md files is an exciting proposal that could enhance personal knowledge management (PKM) systems. This integration would allow seamless interaction between Jan and Obsidian, leveraging AI capabilities for natural language processing, summarization, and contextual information retrieval.

Potential benefits include:
- AI-assisted note creation and organization
- Intelligent search and retrieval within Obsidian vaults
- Context-aware insights and connections between notes

We'd love to hear more from the community about this:
- How many of you use Obsidian and would be interested in a Jan integration?
- What specific features or use cases would you find most valuable?
- Any thoughts on challenges or considerations for this kind of integration?

We're always exploring ways to make Jan more versatile and integrated with popular tools. Your feedback will help us evaluate if this could be a good fit for future development. If there's enough interest (upvotes and comments), we'll seriously consider implementing this feature!

**Comment by 3-ark** [comment]
There are some plugins can support local LLM. Obsidian is a great source for RAG, maybe the best for personal user.

**Reply by Unstuffedspare** [reply]
The only one I know of is BMO, and I wasn't able to get that working (with limited testing).

**Reply by 3-ark** [reply]
"Smart connection" can do this, but right now, the local end seems has a bug, I just submitted a report to the developer. Hope he can fix this soon. but still, general chatting fuction is okay, you can use Jan as backend. Another one called "systemsculpt AI" supports local server with OpenAI format API too. I haven't tried local end with "obsidian copilot", but it updates frequently. Maybe you can try it.

**Reply by Unstuffedspare** [reply]
I will try config Obsidian Copilot. No luck so far, as it appears to automatically connect to Ollama - not sure....

**Reply by jdietl** [reply]
> There are some plugins can support local LLM. Obsidian is a great source for RAG, maybe the best for personal user.

I'm a big fan of Obsidian as a RAG backend, too, particularly in combination with Obsidian Sync: Develop the vault on one machine (possibly with several users), sync to the LLM host automatically, refresh the vector DB (or similar mechanism) automatically at regular intervals.

**Reply by 3-ark** [reply]
Smart connection is my must have plug in now. Besides, recently I found another very good plug-in. Smart composer. Have a try if you use AI for editing. it's pretty new, the developers are still adding new functions, it's way better than other AI plug-in for editing. The local rag so far only support ollama. I have tried Google's free embedding with my sample vault. It did a good job.

**Reply by dan-menlo** [reply]
@jdietl @3-ark @Unstuffedspare Do you know if Obsidian has a Search API that can return pointers to relevant notes? 

Jan's longer-term roadmap is for us to develop "Tools", similar to [OpenAI's Tools](https://platform.openai.com/docs/assistants/tools/function-calling). 

- Tools will retrieve additional context, and pass it into the context window
- e.g. Web Search
- e.g. Website Search (e.g. from a URL that user pastes in)
- e.g. Obsidian Vault Search

If there are no Obsidian APIs for it, it'll likely be a lot more complicated (i.e. the approach that @jdietl outlined above). 

My personal view as an Obsidian user, is that Obsidian Search/Retrieval should make use of the knowledge graph inherent in Obsidian - vs. re-implementing everything from scratch.

**Reply by 3-ark** [reply]
@dan, I think there is one unofficial plug-in like what you need. I know little about coding, so I can't tell. Can you check with it? 
https://github.com/coddingtonbear/obsidian-local-rest-api

Besides, I haven't tried Jan's rag function. I don't know if it supports uploading the whole Obsidian vault into the pipe, it works in this way too. But like you said, it would be much better if you can read the file from API directly.

**Comment by tenten71** [comment]
I love Jan. I love Obsidian. I'd love better integration.
That being said, I don't have the knowledge (yet) of what is involved in making this happen.
Does the model have to be trained on understanding Markdown? Does it have to be trained on Obsidian code and file structure? Does it matter that everyone's vault is setup differently?
Should this or could this be taken a step further and made more general and tell Jan to "look at the files in this folder", tell it what type of files they are (ex: .md, .docx, .pdf, .html), and execute a prompt?
Do we have to specifically program it for Obsidian or can it be programmed to understand any text/code producing program?
Does it have to integrate with the program or can it just integrate with the files?
I'm looking to understand the overall process.

**Reply by vithiri** [reply]
I use Jan and Obsidian as well.

Just being able to add the Obsidian vault folder for RAG and being able to set how often it refreshes (and give and option to manually refresh) would be enough for me.

**Reply by vithiri** [reply]
> I love Jan. I love Obsidian. I'd love better integration.
> That being said, I don't have the knowledge (yet) of what is involved in making this happen.
> Does the model have to be trained on understanding Markdown? Does it have to be trained on Obsidian code and file structure? Does it matter that everyone's vault is setup differently?
> Should this or could this be taken a step further and made more general and tell Jan to "look at the files in this folder", tell it what type of files they are (ex: .md, .docx, .pdf, .html), and execute a prompt?
> Do we have to specifically program it for Obsidian or can it be programmed to understand any text/code producing program?
> Does it have to integrate with the program or can it just integrate with the files?
> I'm looking to understand the overall process.

I've used multiple RAG solutions with Obsidian, and all you need to do is to give it access to the folder, regardless of which model you're using.

The simple instruction "You have read access to the users Obsidian vault." is enough for the model to be moderately helpful.

I usually add "You can find the users personal notes in the Notes folder, and his daily journal notes in the Daily Notes folder. Use these to stay up to date with what the user is doing."

Personalizing the prompt to the way you use Obsidian will improve the outcome, of course, since it helps the model make sense of your personal setup.

**Comment by DartSous** [comment]
I use Obsidian very often and Jan.AI will be very useful there. As an alternative to Notion with its paid AI feature
